File: dispatch/core/Module/ModuleManager.cc
#include "dispatch/core/Module/ModuleManager.hh"

namespace claid {


ModuleManager::ModuleManager(DispatcherClient& dispatcher,
    SharedQueue<DataPackage>& fromModuleDispatcherQueue,
    SharedQueue<DataPackage>& toModuleDispatcherQueue,
    std::shared_ptr<EventTracker> eventTracker) :   dispatcher(dispatcher), 
                                                    fromModuleDispatcherQueue(fromModuleDispatcherQueue), 
                                                    toModuleDispatcherQueue(toModuleDispatcherQueue),
                                                    subscriberPublisher(toModuleDispatcherQueue),
                                                    eventTracker(eventTracker)
{
    Logger::logInfo("ModuleManager constructor event tracker %lu", eventTracker.get());
}

ModuleManager::~ModuleManager()
{
    if(this->restartThread)
    {
        this->restartThread->join();
        this->restartThread = nullptr;
    }
}

absl::Status ModuleManager::instantiateModule(const std::string& moduleId, const std::string& moduleClass)
{
    ModuleFactory& moduleFactory = *ModuleFactory::getInstance();

    if(!moduleFactory.isModuleClassRegistered(moduleClass))
    {
        return absl::NotFoundError(absl::StrCat("ModuleManager: Failed to instantiate Module of class \"", moduleClass, "\" (id: ", moduleId, "\")\n",
        "A Module with this class is not registered to the ModuleFactory."));
    
        // The Middleware always returns all Modules named in the config file.
        // If we cannot load it, we assume another registered runtime can.
        // Therefore, no error.
        //  return true;
    }

    Logger::logInfo("Loaded Module id \"%s\" (class: \"%s\").", moduleId.c_str(), moduleClass.c_str());
    std::unique_ptr<Module> moduleInstance(moduleFactory.getInstanceUntyped(moduleClass));

    moduleInstance->setId(moduleId);
    moduleInstance->setType(moduleClass);
    moduleInstance->setEventTracker(this->eventTracker);
    this->runningModules.insert(make_pair(moduleId, std::move(moduleInstance)));
    return absl::OkStatus();
}

absl::Status ModuleManager::instantiateModules(const ModuleListResponse& moduleList)
{
    for(ModuleListResponse_ModuleDescriptor descriptor : moduleList.descriptors())
    {
        const std::string& moduleId = descriptor.module_id();
        const std::string& moduleClass = descriptor.module_class();

        absl::Status status = this->instantiateModule(moduleId, moduleClass);
        if(!status.ok())
        {
            return status;
        }
    }

    return absl::OkStatus();
}

absl::Status ModuleManager::initializeModules(const ModuleListResponse& moduleList, ChannelSubscriberPublisher& subscriberPublisher)
{
    for(const ModuleListResponse_ModuleDescriptor& descriptor : moduleList.descriptors())
    {
        const std::string& moduleId = descriptor.module_id();
        const std::string& moduleClass = descriptor.module_class();

        const std::string& key = moduleId;
        if(this->runningModules.find(key) == this->runningModules.end())
        {
            return absl::NotFoundError(absl::StrCat(
                "Failed to initialize Module \"" + moduleId + "\" (class: \"" + moduleClass + "\").\n" +
                "The Module was not loaded."
            ));
            
            // The middleware always sends all Modules parsed from the configuration file.
            // If this Module is not running in this runtime, that means we could not instantiate it before.
            // Hence, we assume this Module is supported by another runtime.
            // continue;
        }

        std::unique_ptr<Module>& module = this->runningModules[key];

        
        // This will call the initialize functions of each Module.
        // In the initialize function (and ONLY there), Modules can publish or subscribe Channels.
        // Hence, once all Modules have been initialized, we know the available Channels.
        Logger::logInfo("Calling module.start() for Module \"%s\".", moduleId.c_str());
        
        Properties properties = descriptor.properties();

        if(!module->start(&subscriberPublisher, properties))
        {
            return absl::AbortedError(absl::StrCat(
                "Failed to start Module \"", moduleId.c_str(), "\""
            ));
        }
        Logger::logInfo("Module \"%s\" has started.",  moduleId.c_str());

    }
    return absl::OkStatus();
}

// A Channel is defined by a template package defining source | target | payload.
// For each Module, we have a list of DataPackages defining all Channels published or subscribed by the Module.
// This functions builds a Map<Module, List<DataPackage>>, representing the (data packages for each) channel of the Module.
void ModuleManager::getTemplatePackagesOfModules(std::map<std::string, std::vector<DataPackage>>& moduleChannels)
{
    moduleChannels.clear();
    for(const auto& entry : this->runningModules)
    {
        const std::string& moduleId = entry.first;

        std::vector<DataPackage> templatePackagesForModule = this->subscriberPublisher.getChannelTemplatePackagesForModule(moduleId);
        moduleChannels.insert(make_pair(moduleId, templatePackagesForModule));
    }
}

InitRuntimeRequest ModuleManager::makeInitRuntimeRequest()
{
    InitRuntimeRequest initRuntimeRequest;
    std::map<std::string, std::vector<DataPackage>> moduleTemplatePackages;
    getTemplatePackagesOfModules(moduleTemplatePackages);
    for(const auto& entry : moduleTemplatePackages)
    {
        InitRuntimeRequest_ModuleChannels moduleChannels;

        moduleChannels.set_module_id(entry.first);
        moduleChannels.clear_channel_packets();

        Logger::logInfo("Making init run time request %s, num channels: %d", entry.first.c_str(), entry.second.size());
        for(const DataPackage& package : entry.second)
        {
            *moduleChannels.add_channel_packets() = package;
        }

        *initRuntimeRequest.add_modules() = moduleChannels;
    }   
    initRuntimeRequest.set_runtime(Runtime::RUNTIME_CPP);
    return initRuntimeRequest;
}


absl::Status ModuleManager::start()
{
    Logger::logInfo("CLAID C++ ModuleManager start called");
    if(this->running)
    {
        return absl::AlreadyExistsError("Cannot start C++ Module manager. Start has been called twice.");
    }

    Logger::logInfo("CLAID C++ ModuleManager start called 2");

//    this->dispatcher = std::make_unique<DispatcherClient>(socketPath, fromModuleDispatcherQueue, toModuleDispatcherQueue, );
    std::unique_ptr<ModuleListResponse> moduleList = this->dispatcher.getModuleList();
    if(moduleList == nullptr)
    {
        return absl::AbortedError("Failed to receive ModuleListResponse from middleware.");
    }
        Logger::logInfo("CLAID C++ ModuleManager start called 3");
        Logger::logInfo("%s", messageToString(*moduleList).c_str());

    absl::Status status;
    Logger::logInfo("CLAID C++ ModuleManager start called 4");

    status = instantiateModules(*moduleList);
    if(!status.ok())
    {
        return status;
    }
    Logger::logInfo("CLAID C++ ModuleManager start called 5");


    status = initializeModules(*moduleList, subscriberPublisher);
    if(!status.ok())
    {
        return status;
    }


    InitRuntimeRequest initRuntimeRequest = makeInitRuntimeRequest();

    grpc::Status dispatcherStatus = this->dispatcher.startRuntime(initRuntimeRequest);
    if(!dispatcherStatus.ok())
    {
        return absl::Status(static_cast<absl::StatusCode>(dispatcherStatus.error_code()),
                       dispatcherStatus.error_message());
    }

    this->running = true;


    this->fromModuleDispatcherReaderThread = std::thread(&ModuleManager::readFromModulesDispatcher, this);

    Logger::logInfo("CLAID C++ runtime started successfully!");

    return absl::OkStatus();
}

void ModuleManager::shutdownModules()
{
    Logger::logInfo("ModuleManager shutting down Modules.");

    for(const auto& entry : this->runningModules)
    {
        const std::string& moduleName = entry.first;
        Module& module = *entry.second;
        Logger::logInfo("ModuleManager shutting down Module \"%s\".", moduleName.c_str());
        module.shutdown();
    }
    this->runningModules.clear();
}

void ModuleManager::stop()
{
    if(!this->running)
    {
        return;
    }

    Logger::logInfo("ModuleManager waiting for reader thread to stop.");

    this->shutdownModules();
    Logger::logInfo("All Modules shutdown in C++ runtime.");

    this->running = false;
    this->fromModuleDispatcherQueue.interruptOnce();
    this->fromModuleDispatcherReaderThread.join();
    Logger::logInfo("ModuleManager has stopped.");
    // Reset the subscriber publisher
    this->subscriberPublisher.reset();
}


void ModuleManager::readFromModulesDispatcher()
{
    while(this->running)
    {
        std::shared_ptr<DataPackage> pkt = fromModuleDispatcherQueue.interruptable_pop_front();

        if (!pkt) 
        {
            Logger::logInfo("ModuleManager received nullptr package from ModuleDispatcher");
            continue;
        }

        onPackageReceivedFromModulesDispatcher(pkt);
    }
}

void ModuleManager::onPackageReceivedFromModulesDispatcher(std::shared_ptr<DataPackage> dataPackage)
{
    /*String[] hostAndModule = splitHostModule(dataPackage.getTargetModule());

    if(hostAndModule == null)
    {
        Logger.logError("Java Runtime received DataPackage with invalid address \"" + dataPackage.getTargetModule() + "\".\n"
        + "Unable to split the address into host:module");
    }*/

    if(dataPackage->has_control_val())
    {
        Logger::logInfo("ModuleManager received DataPackage with controlVal %d (%s)", dataPackage->control_val().ctrl_type(), CtrlType_Name(dataPackage->control_val().ctrl_type()).c_str());
        handlePackageWithControlVal(dataPackage);
        return;
    }

    /*String hostName = hostAndModule[0];
    String moduleId = hostAndModule[1];*/

    const std::string& channelName = dataPackage->channel();
    const std::string& moduleId = dataPackage->target_module();

    Logger::logInfo("ModuleManager received package with target for Module \"%s\" on Channel \"%s\"",
                    moduleId.c_str(), channelName.c_str());


    if(!subscriberPublisher.isDataPackageCompatibleWithChannel(*dataPackage, moduleId))
    {
        const std::string payloadCaseName = dataPackagePayloadCaseToString(*dataPackage);
        Logger::logInfo("ModuleManager received package with target for Module \"%s\" on Channel \"%s\",\n"
            "however the data type of payload of the package did not match the data type of the Channel.\n"
            "Expected payload type \"%s\" but got \"%s\".",
            moduleId.c_str(), channelName.c_str(), payloadCaseName.c_str(), payloadCaseName.c_str());

        return;
    }

    std::vector<std::shared_ptr<AbstractSubscriber>> subscriberList 
        = this->subscriberPublisher.getSubscriberInstancesOfModule(channelName, moduleId);

    Logger::logInfo("ModuleManager invoking subscribers. Found %d subscribers.", subscriberList.size());
    for(std::shared_ptr<AbstractSubscriber> subscriber : subscriberList)
    {
        Logger::logInfo("Invoking subscriber %lu", subscriber.get());
        subscriber->onNewData(dataPackage);
    
    }
    Logger::logInfo("ModuleManager done invoking subscribers.", subscriberList.size());

}

void ModuleManager::handlePackageWithControlVal(std::shared_ptr<DataPackage> package)
{
    switch(package->control_val().ctrl_type())
    {
        case CtrlType::CTRL_CONNECTED_TO_REMOTE_SERVER:
        {
            for(auto& modulesEntry : this->runningModules)
            {
                modulesEntry.second->notifyConnectedToRemoteServer();
            }
            break;
        }
        case CtrlType::CTRL_DISCONNECTED_FROM_REMOTE_SERVER:
        {
            for(auto& modulesEntry : this->runningModules)
            {
                modulesEntry.second->notifyDisconnectedFromRemoteServer();
            }
            break;
        }
        case CtrlType::CTRL_UNLOAD_MODULES:
        {   
            Logger::logInfo("ModuleManager received ControlPackage with code CTRL_UNLOAD_MODULES");
            Logger::logInfo("Unloading modules!");
            this->shutdownModules();
            std::shared_ptr<DataPackage> response = std::make_shared<DataPackage>();
            ControlPackage& ctrlPackage = *response->mutable_control_val();
            
            ctrlPackage.set_ctrl_type(CtrlType::CTRL_UNLOAD_MODULES_DONE);
            ctrlPackage.set_runtime(Runtime::RUNTIME_CPP);
            response->set_source_host(package->target_host());
            response->set_target_host(package->source_host());
            toModuleDispatcherQueue.push_back(response);
            Logger::logInfo("Unloading modules done and acknowledgement will be sent out");
            break;
        }
        case CtrlType::CTRL_RESTART_RUNTIME:
        {   
            // Need to switch to separate thread, because stop would call fromModuleDispatcherReaderThread,
            // which is the same thread as the one we are currently in.
            if(this->restartThread)
            {
                this->restartThread->join();
            }
            this->restartThread = std::make_unique<std::thread>(&ModuleManager::restart, this);

            restartControlPackage = *package.get();
            
            
            break;
        }
        case CtrlType::CTRL_PAUSE_MODULE:
        {
            Logger::logWarning("Received pauise module %s", package->target_module().c_str());
            const std::string& targetModule = package->target_module();

            auto it = this->runningModules.find(targetModule);
            if(it == this->runningModules.end())
            {
                // Logger::logError("Got request to pause Module \"%s\", but Module with that id could not be found. Is it running?");
                return;
            }
                        Logger::logWarning("Received pauise module  2");

            it->second->pauseModule();
                                    Logger::logWarning("Received pauise module  3");
            break;
        }
        case CtrlType::CTRL_UNPAUSE_MODULE:
        {
            const std::string& targetModule = package->target_module();

            auto it = this->runningModules.find(targetModule);
            if(it == this->runningModules.end())
            {
                // Logger::logError("Got request to pause Module \"%s\", but Module with that id could not be found. Is it running?");
                return;
            }
            it->second->resumeModule();
            break;
        }
        case CtrlType::CTRL_ADJUST_POWER_PROFILE:
        {
            const std::string& targetModule = package->target_module();

            auto it = this->runningModules.find(targetModule);
            if(it == this->runningModules.end())
            {
                // Logger::logError("Got request to pause Module \"%s\", but Module with that id could not be found. Is it running?");
                return;
            }
            it->second->adjustPowerProfile(package->control_val().power_profile());
            break;
        }
        default:
        {
            Logger::logWarning("ModuleManager received package with unsupported control val %d", package->control_val().ctrl_type());
        }
        break;
    }
}

void ModuleManager::restart()
{
    Logger::logInfo("ModuleManager received ControlPackage with code CTRL_RESTART_RUNTIME");
    Logger::logInfo("ModuleManager stopping");
    this->stop();
    Logger::logInfo("ModuleManager stopping dispatcher");
    this->dispatcher.shutdown();
    Logger::logInfo("ModuleManager restarting");
    absl::Status status = this->start();
    if(!status.ok())
    {
        Logger::logError("Failed to restart: %s", status.ToString().c_str());
        return;
    }
    Logger::logInfo("...aaaaaandddd done!");

    std::shared_ptr<DataPackage> response = std::make_shared<DataPackage>();
    ControlPackage& ctrlPackage = *response->mutable_control_val();

    ctrlPackage.set_ctrl_type(CtrlType::CTRL_RESTART_RUNTIME_DONE);
    ctrlPackage.set_runtime(Runtime::RUNTIME_CPP);
    response->set_source_host(restartControlPackage.target_host());
    response->set_target_host(restartControlPackage.source_host());
    toModuleDispatcherQueue.push_back(response);
    
}

}

File: dispatch/python/local_dispatching/module_manager.py
from module.module_factory import ModuleFactory
from logger.logger import Logger
from module.channel_subscriber_publisher import ChannelSubscriberPublisher
from module.module_annotator import ModuleAnnotator
from module.thread_safe_channel import ThreadSafeChannel
from local_dispatching.module_injector import ModuleInjector
from dispatch.proto.claidservice_pb2 import * 

from threading import Thread
import time

class ModuleManager():

    def __init__(self, module_dispatcher, module_factory, modules_injection_storage_path = None):
        self.__module_dispatcher = module_dispatcher
        self.__module_factory = module_factory
        # In Python, it is possible to add Modules at runtime.

        if modules_injection_storage_path is not None:
            self.__module_injector = ModuleInjector(modules_injection_storage_path, self.__module_factory)
        self.__running_modules = dict()    

        # self.__from_module_dispatcher_queue = ThreadSafeChannel()
        self.__to_module_dispatcher_queue = module_dispatcher.get_to_dispatcher_queue()
        
        self.__channel_subscriber_publisher = ChannelSubscriberPublisher(self.__to_module_dispatcher_queue)

        self.__running = False
        self.__from_module_dispatcher_queue_closed = False

        self.__module__anotations_updated = False


    def instantiate_module(self, module_id, module_class):

        if not self.__module_factory.is_module_class_registered(module_class):
            Logger.log_error(f"ModuleManager: Failed to instantiate Module of class \"{module_class}\" (id: {module_id}\")\n"
                  f"A Module with this class is not registered to the ModuleFactory.")
            return False

        Logger.log_info(f"Loaded Module id \"{module_id}\" (class: \"{module_class}\").")
        module = self.__module_factory.get_instance(module_class, module_id)
        self.__running_modules[module_id] = module
        print(self.__running_modules)
        return True

    def instantiate_modules(self, module_list):
        print("bla ", module_list.descriptors)
        for descriptor in module_list.descriptors:
            module_id = descriptor.module_id
            module_class = descriptor.module_class

            if not self.instantiate_module(module_id, module_class):
                Logger.log_error(f"Failed to instantiate Module \"{module_id}\" (class: \"{module_class}\").\n"
                      f"The Module class was not registered to the ModuleFactory.")
                return False
        return True

    def initialize_modules(self, module_list, subscriber_publisher):
        for descriptor in module_list.descriptors:
            module_id = descriptor.module_id
            module_class = descriptor.module_class

            key = module_id
            if key not in self.__running_modules:
                Logger.log_error(f"Failed to initialize Module \"{module_id}\" (class: \"{module_class}\").\n"
                      f"The Module was not loaded.")
                return False

            module = self.__running_modules[key]

            Logger.log_info(f"Calling module.start() for Module \"{module.get_id()}\".")
            module.start(subscriber_publisher, descriptor.properties)
            Logger.log_info(f"Module \"{module.get_id()}\" has started.")

        return True

    def get_template_packages_of_modules(self):
        module_channels = {}
        for module_id in self.__running_modules.keys():
            template_packages_for_module = self.__channel_subscriber_publisher.get_channel_template_packages_for_module(module_id)
            module_channels[module_id] = template_packages_for_module
        return module_channels

    def start(self):
        print("ModuleManager1")
        print(self.__module_factory.get_registered_module_classes())
        print("ModuleManager2")

        registered_module_classes = self.__module_factory.get_registered_module_classes()
        module_annotations = dict()
        print("ModuleManager3")

        for registered_module_class in registered_module_classes:

            module_annotator = ModuleAnnotator(registered_module_class)
            has_annotate_module_function = self.__module_factory.get_module_annotation(registered_module_class, module_annotator)
       

            if(has_annotate_module_function):
                module_annotations[registered_module_class] = module_annotator.get_annotations()
        print("ModuleManager4")

        module_list =  self.__module_dispatcher.get_module_list(registered_module_classes, module_annotations)
        Logger.log_info(f"Received ModuleListResponse: {module_list}")
        if not self.instantiate_modules(module_list):
            Logger.log_fatal("ModuleDispatcher: Failed to instantiate Modules.")
            return False

        print("ModuleManager5")

        if not self.initialize_modules(module_list, self.__channel_subscriber_publisher):
            Logger.log_fatal("Failed to initialize Modules.")
            return False
        print("ModuleManager6")

        example_packages_of_modules = self.get_template_packages_of_modules()
        if not  self.__module_dispatcher.init_runtime(example_packages_of_modules):
            Logger.log_fatal("Failed to initialize runtime.")
            return False

        self.__from_module_dispatcher_queue_closed = False
        if not self.__module_dispatcher.send_receive_packages():
            Logger.log_fatal("Failed to set up input and output streams with middleware.")
            return False

        self.__from_module_dispatcher_queue = self.__module_dispatcher.get_from_dispatcher_queue()

        self.running = True
        self.read_from_module_dispatcher_thread = Thread(target=self.read_from_module_dispatcher)
        self.read_from_module_dispatcher_thread.start()

        return True
    
    def shutdown_modules(self):
        for module_name, module in self.__running_modules.items():
            print("Shutting down ", module_name)
            module.shutdown()
            print("Module has shutdown")
        
        self.__running_modules.clear()

    def stop(self):
         
        self.running = False
        self.__from_module_dispatcher_queue.cancel()
        self.read_from_module_dispatcher_thread.join()
        
        self.__channel_subscriber_publisher.reset()
    
    # def start_test(self):
    #     self.start()
    #     # io.ensure_future(self.start())

       
    #     # try:
    #     #     # Run the event loop
    #     #     io.get_event_loop().run_forever()
    #     # except KeyboardInterrupt:
    #     #     pass

    def split_host_module(self, addr):
        host_and_module = addr.split(":")

        if len(host_and_module) != 2:
            return None

        return host_and_module

    def on_data_package_received_from_module_dispatcher(self, data_package):
        if data_package.HasField('control_val'):
            self.handle_package_with_control_val(data_package)
            return

        channel_name = data_package.channel
        module_id = data_package.target_module

        Logger.log_info(f"ModuleManager received package with target for Module \"{module_id}\" on Channel \"{channel_name}\"")

        if self.__channel_subscriber_publisher is None:
            Logger.log_error("ModuleManager received DataPackage, however SubscriberPublisher is Null.")
            return

        if not self.__channel_subscriber_publisher.is_data_package_compatible_with_channel(data_package, module_id):
            payload_case = data_package.WhichOneof("payload_oneof")
            Logger.log_info(f"ModuleManager received package with target for Module \"{module_id}\" on Channel \"{channel_name}\",\n"
                            f"however the data type of payload of the package did not match the data type of the Channel.\n"
                            f"Expected payload type \"{self.__channel_subscriber_publisher.get_payload_case_of_channel(channel_name).name}\" but got \"{payload_case}")
            return

        subscriber_list = self.__channel_subscriber_publisher.get_subscriber_instances_of_module(channel_name, module_id)

        if subscriber_list is None:
            Logger.log_info(f"ModuleManager received package with target for Module \"{module_id}\" on Channel \"{channel_name}\",\n"
                            f"however a Subscriber of the Module for this Channel was not found. The Module has no Subscriber for this Channel.")
        for subscriber in subscriber_list:
            subscriber.on_new_data(data_package)

    def handle_package_with_control_val(self, packet):
        Logger.log_info("PYTHON HANDLE PACKAGE")
        if packet.control_val.ctrl_type == CtrlType.CTRL_CONNECTED_TO_REMOTE_SERVER:
            for module_id, module in self.__running_modules.items():
                module.notify_connected_to_remote_server()
        elif packet.control_val.ctrl_type == CtrlType.CTRL_DISCONNECTED_FROM_REMOTE_SERVER:
            for module_id, module in self.__running_modules.items():
                module.notify_disconnected_from_remote_server()
        elif packet.control_val.ctrl_type == CtrlType.CTRL_UNLOAD_MODULES:
            self.shutdown_modules()

            Logger.log_info("Python ModuleManager received CTRL_UNLOAD_MODULES")

            response = DataPackage()
            ctrl_package = response.control_val

            ctrl_package.ctrl_type = CtrlType.CTRL_UNLOAD_MODULES_DONE
            ctrl_package.runtime = Runtime.RUNTIME_PYTHON
            response.source_host = packet.target_host
            response.target_host = packet.source_host

            self.__to_module_dispatcher_queue.put(response)

            Logger.log_info("Unloading Modules done")
        elif packet.control_val.ctrl_type == CtrlType.CTRL_RESTART_RUNTIME:
            
            Logger.log_info("Python ModuleManager received CTRL_RESTART_RUNTIME")
            self.restart_thread = Thread(target=self.restart)
            self.restart_thread.start()

            self.__restart_control_package = packet
        elif packet.control_val.ctrl_type == CtrlType.CTRL_REQUEST_MODULE_ANNOTATIONS_RESPONSE:
            self.__module_annotations = packet.control_val.module_annotations
            self.__module__anotations_updated = True
        elif packet.control_val.ctrl_type == CtrlType.CTRL_ON_NEW_CONFIG_PAYLOAD_DATA:
            payload = packet.control_val.config_upload_payload
            
            self.inject_modules_from_config_upload_payload(payload)

        else:
            Logger.log_warning(f"ModuleManager received package with unsupported control val {packet.control_val.ctrl_type}")

    def read_from_module_dispatcher(self):
        while self.running:
            print("on read from module dispatcher")
            try:
                for data_package in self.__from_module_dispatcher_queue:
                    if data_package is not None:
                        self.on_data_package_received_from_module_dispatcher(data_package)
                    else:
                        pass
            except Exception as e:
                print(e)
        
        self.__from_module_dispatcher_queue_closed = True


    

    def restart(self):
        Logger.log_info("Stopping ModuleManager")
        self.stop()
        Logger.log_info("Stopping ModuleDispatcher")
        self.__module_dispatcher.shutdown()

        print("Waiting42")
        while not self.__from_module_dispatcher_queue_closed:
            print("Waiting")
        time.sleep(2)
        Logger.log_info("Restarting ModuleManager and ModuleDispatcher")
        self.start()

        response = DataPackage()
        ctrl_package = response.control_val

        ctrl_package.ctrl_type = CtrlType.CTRL_RESTART_RUNTIME_DONE
        ctrl_package.runtime = Runtime.RUNTIME_PYTHON
        response.source_host = self.__restart_control_package.target_host
        response.target_host = self.__restart_control_package.source_host

        self.__to_module_dispatcher_queue.put(response)

    def update_module_annotations(self):
        self.__module__anotations_updated = False
        
        package = DataPackage()
        ctrl_package = package.control_val

        ctrl_package.ctrl_type = CtrlType.CTRL_REQUEST_MODULE_ANNOTATIONS
        ctrl_package.runtime = Runtime.RUNTIME_PYTHON
   

        self.__to_module_dispatcher_queue.put(package)

    def are_module_annotations_updated(self):
        return self.__module__anotations_updated
    
    def get_module_annotations(self):
        return self.__module_annotations
    
    # dict[str, List]
    # dict(code, module_names)
    def inject_new_modules(self, module_descrption: dict):

        if self.__module_injector is None:
            return False

        for code_name in module_descrption:
            code, module_names = module_descrption[code_name]
            if not self.__module_injector.inject_claid_modules_from_python_file(code_name, code, module_names):
                return False
            
        
        self.__module_factory = self.__module_injector.rebuild_get_module_factory()
        self.__module_factory.print_registered_modules()

        return True


    def inject_modules_from_config_upload_payload(self, payload):
        